presto-accumulo/src/main/java/com/facebook/presto/accumulo/AccumuloClient.java

```diff
@@ -0,0 +1,1140 @@
+                    if (m.getValue().getRight().equals(qual)) {
+                        tryAgain = true;
+                        break;
+                    }
+                }
+
+                if (!tryAgain) {
+                    mapping.put(column.getName(), Pair.of(fam, qual));
+                }
+            }
+            while (tryAgain);
+        }
+        return mapping;
+    }
+
+    /**
+     * Creates the index tables from the given Accumulo table. No op if
+     * {@link AccumuloTable#isIndexed()} is false.
+     *
+     * @param table Table to create index tables
+     */
+    private void createIndexTables(AccumuloTable table)
+    {
+        try {
+            // If our table is indexed, create the index as well
+            if (table.isIndexed()) {
+                // Create index table and set the locality groups
+                Map<String, Set<Text>> indexGroups = Indexer.getLocalityGroups(table);
+                conn.tableOperations().create(table.getIndexTableName());
+                conn.tableOperations().setLocalityGroups(table.getIndexTableName(), indexGroups);
+
+                // Create index metrics table, attach iterators, and set locality groups
+                conn.tableOperations().create(table.getMetricsTableName());
+                conn.tableOperations().setLocalityGroups(table.getMetricsTableName(), indexGroups);
+                for (IteratorSetting s : Indexer.getMetricIterators(table)) {
+                    conn.tableOperations().attachIterator(table.getMetricsTableName(), s);
+                }
+            }
+        }
+        catch (Exception e) {
+            dropTable(table);
+            throw new PrestoException(INTERNAL_ERROR,
+                    "Accumulo error when creating index tables, check coordinator logs for more detail",
+                    e);
+        }
+    }
+
+    /**
+     * Drops the table metadata from Presto. If this table is internal, the Accumulo
+     * tables are also deleted.
+     *
+     * @param table The Accumulo table
+     */
+    public void dropTable(AccumuloTable table)
+    {
+        SchemaTableName stName = new SchemaTableName(table.getSchema(), table.getTable());
+        String tn = table.getFullTableName();
+
+        // Remove the table metadata from Presto
+        metaManager.deleteTableMetadata(stName);
+
+        // If the table is external, we won't clean up the Accumulo tables
+        if (table.isExternal()) {
+            return;
+        }
+
+        try {
+            // delete the table
+            conn.tableOperations().delete(tn);
+
+            // and delete our index tables
+            if (table.isIndexed()) {
+                conn.tableOperations().delete(Indexer.getIndexTableName(stName));
+                conn.tableOperations().delete(Indexer.getMetricsTableName(stName));
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(INTERNAL_ERROR, "Failed to delete table",
```

Comment:
I was thinking about the whole workflow for situations like these and am curious on your thoughts.  This came up as I was refactoring the table creation process into something a bit easier to digest.  There are a lot of steps where, at any point, some Accumulo or ZooKeeper error could occur -- either the system fails or, say, a table is created by an external process in between the time I check for table existence and it is created:

1. All sorts of validation steps to ensure the table is correct and can be created
2. Create metadata in ZooKeeper
3. Create namespace
4. Create table
5. Set locality groups on main table
6. Create index table
7. Set locality groups on index table
8. Create metrics table
9. Set iterators on metrics table

The code for the create table case is getting pretty ugly, where if any of the above things occurs, the connector attempts to rollback the operation.  However, if an error occurred, it will likely occur again (say, in the case that Accumulo went down) and the rollback will also fail, leaving a discrepancy between what the connector thinks happened and what actually happened.  This would then require some documentation for the end-users to demonstrate to them how to clean up the state by hand once the services are restored.

Now, I am leaning towards a flow more akin to:
1. Validate everything is good to perform the operation
2. Perform the operation
3. If an Accumulo/ZK error occurs, throw an appropriate error and direct users to documentation to resolve the issue (vs. trying to rollback and likely failing)

To answer your question -- yes, it is possible that the first delete will pass and the second will fail.  This will result in a zombie table in Accumulo, and if a user tries to create an indexed table of the same name, they'll get an error saying the Accumulo table already exists.  Theyw ill then have to go delete the table by hand before they can create the new table.  By leaning towards this approach in the first place, it'll make the code easier to read and maintain, while also giving the users appropriate documentation to resolve any issues caused by unexpected Accumulo/ZooKeeper errors.  Do you think this is a reasonable approach to error handling?